### PR TITLE
Rely on Logical Dimensions to Constrain the Size of Merged Dimensions

### PR DESCRIPTION
**Pull request by ulysseB**
Changes how we ensure merged dimensions have the same size. We used to call `is_compatible_with` method on dimension sizes, but this method won't work when we enable strip-mining: we can't test the equality on `PartialSize` as they will only be partially specified. Instead we ensure logical dimensions have the same size and that the tiling factor is the same.

We only make changes for dimensions with a dynamic size. Dimensions with a static size where already covered.

This PR also fixes a small bug in TelamonGen. Some code was inside a loop when it should not have been. The result was that it was not run when the loop had 0 iterations and generated invalid code in that case.

### CODE REVIEW

File: telamon-gen/src/ir/mod.rs [hunk 1]
```diff
@@ -260,10 +260,9 @@ impl IrDesc {
         (arg_foralls, set_constraints, adaptator)
     }
 
-    /// Generates the list of sets to iterate and to constraints to iterate on the given
-    /// environement, but from the point of view of the given choice instance. The new
-    /// foralls iterating on current arguments are returned in a separate list than the
-    /// ones issued from foralls.
+    /// Generates the foralls and the set constraints to iterate on the given environement,
```
Comment (nicoTolly) [inline telamon-gen/src/ir/mod.rs]: Environment (typo)

File: telamon-gen/src/ir/mod.rs [hunk 2]
```diff
@@ -260,10 +260,9 @@ impl IrDesc {
...
+    /// from the point of view of the given choice. Returns the foralls issued from arguments
+    /// in a different list than the foralls issued from foralls in the original environement.
```
Comment (nicoTolly) [inline telamon-gen/src/ir/mod.rs]: Same

**Comment by ulysseB** [comment]
bors retry

**Pull request merged**